[DOCS-3411] Add new security notification and variables docs (#14881)
DataDog/documentation: content/en/security_platform/notifications/variables.md

---
title: Variables
kind: documentation
further_reading:
- link: "/security_platform/detection_rules/"
  tag: "Documentation"
  text: "Explore security detection rules"
- link: "/security_platform/notifications/"
  tag: "Documentation"
  text: "Learn more about Security Platform notifications"
---

## Overview

> When [creating a new detection rule or modifying an existing one][1], use [template variables](#template-variables), such as attributes and signal tags, and [conditional variables](#conditional-variables) to customize a rule’s notification message. When a signal is generated from the rule, the variables are populated with values related to that signal. 

Reviewer: ```suggestion
When [creating a new detection rule or modifying an existing one][1], use [template variables](#template-variables) (such as attributes and signal tags) and [conditional variables](#conditional-variables) to customize a rule’s notification message. When a signal is generated from the rule, the variables are populated with values related to that signal. 
```

| `{{last_seen_epoch}}` | Time the signal was most recently triggered, in milliseconds, since midnight, January 1, 1970.|
| `{{rule_name}}`       | Name of the associated rule.                                                                  |
| `{{case_name}}`       | Name of the triggering rule case.                                                             |

### Dynamic links

Use template variables to dynamically link to a related resource for your investigation. 

For example, if a signal detects a suspicious user login, use `{{@user.id}}` to create a dynamic link to another resource:

```
* [Investigate user in the authentication dashboard](https://app.datadoghq.com/example/integration/security-monitoring---authentication-events?tpl_var_username={{@usr.id}})
```

Or, if a signal is tagged with a specific service, use the `{{@service}}` variable to create a dynamic link:

```
* [Investigate service in the services dashboard](https://app.datadoghq.com/example/integration/application-security---service-events?tpl_var_service={{@service}})
```

### Evaluation of numerical values

> For template variables that return numerical values, use `eval` to perform mathematical operations or change the value’s format. For full details, see [Template Variable Evaluation][2].

Reviewer: ```suggestion
For template variables that return numerical values, use `eval` to perform mathematical operations or change the value’s format. For more information, see [Template Variable Evaluation][2].
```

### Epoch

Epoch template variables create a human-readable string or math-friendly number within a notification. For example, use values such as `first_seen`, `last_seen`, or `timestamp` (in milliseconds) within a function to receive a readable string in a notification. For example:

```
{{eval "first_seen_epoch-15*60*1000"}}
```

For more information on the `eval` function, see [Template Variable Evaluation][2].

### Local time

> Use the local_time function to add another date in your notification in the time zone of your choice. This function transforms a date into its local time: `{{local_time 'time_variable' 'timezone'}}`. 

Reviewer: ```suggestion
Use the `local_time` function to add another date in your notification in the time zone of your choice. This function transforms a date into its local time: `{{local_time 'time_variable' 'timezone'}}`. 
```

For example, to add the last triggered time of the signal in the Tokyo time zone in your notification, include the following in the notification message:

```
{{local_time 'last_triggered_at' 'Asia/Tokyo'}}
```

> The result is displayed in the ISO 8601 format: `yyyy-MM-dd HH:mm:ss±HH:mm`, for example, `2021-05-31 23:43:27+09:00`. See the [list of TZ database time zones][3], specifically the `TZ database name` column to see the list of available time zone values.

Reviewer: ```suggestion
The result is displayed in the ISO 8601 format: `yyyy-MM-dd HH:mm:ss±HH:mm`, for example, `2021-05-31 23:43:27+09:00`. See the [list of TZ database time zones][3], specifically the `TZ database name` column, to see the list of available time zone values.
```

## Attribute variables

Use attribute variables to customize signal notifications with specific information about the triggered signal. 

> To see a signal’s list of event attributes, click **JSON** at the bottom of the Overview tab in the signal’s side panel. Use the following syntax to add these event attributes in your rule notifications: `{{@attribute}}`. To access inner keys of the event attributes, use JSON dot notation, for example, `{{@attribute.inner_key}})`.

Reviewer: ```suggestion
To see a signal’s list of event attributes, click **JSON** at the bottom of the **Overview** tab in the signal’s side panel. Use the following syntax to add these event attributes in your rule notifications: `{{@attribute}}`. To access inner keys of the event attributes, use JSON dot notation, for example, `{{@attribute.inner_key}})`.
```

{{< tabs >}}
{{% tab "Cloud SIEM" %}}

```json
{
  "network": {
    "client": {
      "ip": "1.2.3.4"
    }
  },
  "usr": {
    "id": "user@domain.com"
  },
  "evt": {
    "category": "authentication",
    "outcome": "success"
  },
  "used_mfa": "false"
}
```

If you use the following in the “Say what’s happening” section:

```
{{@usr.id}} just logged in without MFA from {{@network.client.ip}}.
```
This is what the notification message looks like:

```
user@domain.com just logged in without MFA from 1.2.3.4.
```

{{% /tab %}}

{{% tab "Application Security Monitoring" %}}

```json
{
  "attributes":{
    "title":"Security scanner detected",
    "http":{
      "url":"http://www.example.com"
    },
    "rule":{
      "detectionMethod":"threshold",
      "name":"Your rule name"
    },
    "events_matched":2,
    "first_seen":"2022-01-26T13:23:33.000Z",
    "last_seen":"2022-01-27T04:01:57.000Z"
  },
  "groupByPaths":[
    "service"
  ]
}
```

If you use the following in the Say What's Happening section:

```
Real routes targeted for {{@service}}.
```

The notification shows the service name in the message as follows:

```
Real routes targeted for your_service_name.
```

> {{% /tab %}}

Reviewer 2: We need to swap the content of the tabs

Author: Hey @btoueg! The content was taken directly from the specific product detection rule pages, which was what PMs had put in:

From Cloud SIEM's log detection rule page: https://docs.datadoghq.com/security_platform/cloud_siem/log_detection_rules?tab=threshold#say-whats-happening 

From ASM's custom detection page: https://docs.datadoghq.com/security_platform/application_security/custom_rules/#template-variables

Reviewer 2: Ok my bad, I'll see if I can provide better suggestion later

{{< /tabs >}}

### More examples

Use `{{@network.client.ip}}` to display the IP address(es) associated with the signal.

> If a security rule detects a user logging in from an IP address known to be malicious, use the template variables `{{@usr.id}}` and `{{@network.client.ip}}` to say which user and IP address triggered the signal. For example:

Reviewer: Is this wording "say" or "see?"

```suggestion
If a security rule detects a user logging in from an IP address known to be malicious, use the template variables `{{@usr.id}}` and `{{@network.client.ip}}` to see which user and IP address triggered the signal. For example:
```

```
The user {{@usr.id}} just successfully authenticated from {{@network.client.ip}} which is a known malicious IP address.
```
## Tag variables

Use the following syntax to add a tag variable to your rule’s notification message: `{{tag_name}}`. 

For tags following the `key:value` syntax, use the variable: `{{key.name}}`. This renders the value associated with the key in the notification. For example, if a signal has the tag key `region`, use the variable `{{region.name}}` in your notification message.

> **Note:** There is no need to use `@` to access the tag value.

Reviewer: ```suggestion
There is no need to use `@` to access the tag value.
```
